R/Plot_synergy.R

```diff
@@ -1,328 +1,330 @@
+# method used to compute them.
+.ComputePlotTitle <- function(scores.dose, row.range, col.range, data.method) {
+  # The first row and column are removed no matter what,
+  # presumably (? TODO/REVIEW) on the assumptions that:
+  #  1. For each drug, the first recorded effect is for dose=0.
+  #  2. When dose=0 for either drug, the synergy score is also 0.
+  return(paste("Average synergy: ",
+               round(mean(scores.dose[.SliceAtLeastFirst(row.range),
+                                      .SliceAtLeastFirst(col.range)],
+                          na.rm = TRUE),
+                     3),
+               " (",data.method,")",
+               sep = ""))
+}
+
+# Convert 'val' to a slice.  However, guarantee that the slice starts at
+# 2 or higher.  So, both for val=c(1,5) and val=c(2,5), the function should
+# return 2:5.  The first two elements of list 'val' are taken to be the slice
+# endpoints; any other elements are ignored.  If 'val' is NULL, then the
+# returned slice is '-1' -- 'remove the first element'.
+.SliceAtLeastFirst <- function(val) {
+  return(if(is.null(val)) -1 else max(val[1],2):val[2])
+}
 
-  if(!save.file) {
-    for(i in num.pairs) {
-      dev.new(noRStudioGD = TRUE)
-      replayPlot(plots[[i]])
-    }
+# Returns a slice of 'val' specified by 'range', as numeric.
+# If 'range' is null, 'val' is wholly returned.  The first two elements of
+# 'range' are taken to be slice endpoints; any other values are ignored.
+# The desired portion of 'val' is converted to numeric form, then returned.  An
+# error occurs if any element of the selected slice of 'val' cannot be converted
+# to a number.
+.NumericApplyRange <- function(val, range) {
+  return(as.numeric(if(is.null(range)) val else val[range[1]:range[2]]))
+}
+
+# Returns a 2D slice of 'data' specified by 'row.range' and 'col.range'.  If
+# either range is NULL, the data for the corresponding axis of 'data' is
+# returned in full.  The row.range and col.range inputs are each treated as
+# lists whose first two elements are range endpoints, and whose remaining
+# elements (if any) are ignored.
+.ApplyRanges <- function(data, row.range, col.range) {
+  .RangeWithDefault <- function(range, default) {
+    # TODO/REVIEW: The code checked into GitHub at the time this function was
+    # written modified each element of each range with the transformation x ->
+    # (x-1)*5. It had no examples that used ranges.  The transform seemed out of
+    # place and is omitted here.
+    return(if(is.null(range)) default else range[1]:range[2])
   }
+  # Default value is 'remove the element just after the end of the list',
+  # for each dimension of the array (a no-op that permits cleaner syntax).
+  return(data[.RangeWithDefault(row.range, default=-(nrow(data)+1)),
+              .RangeWithDefault(col.range, default=-(ncol(data)+1))])
+}
+
+# TODO/REVIEW: The original logic checked into GitHub at the time this function
```

Comment:
Please pay special attention to this TODO/REVIEW comment.  It should get cleaned up one way or another before the changes are submitted.